tooling/nargo_cli/src/cli/compile_cmd.rs

```diff
@@ -54,28 +62,77 @@ pub(crate) fn run(
         selection,
         Some(NOIR_ARTIFACT_VERSION_STRING.to_owned()),
     )?;
-    let circuit_dir = workspace.target_directory_path();
 
+    if args.compile_options.expression_width.is_none() {
+        args.compile_options.expression_width = Some(backend.get_backend_info_or_default());
+    };
+
+    if args.watch {
+        watch_workspace(&workspace, &args.compile_options)
+            .map_err(|err| CliError::Generic(err.to_string()))?;
+    } else {
+        compile_workspace_full(&workspace, &args.compile_options)?;
+    }
+
+    Ok(())
+}
+
+fn watch_workspace(workspace: &Workspace, compile_options: &CompileOptions) -> notify::Result<()> {
+    let (tx, rx) = std::sync::mpsc::channel();
+
+    // No specific tickrate, max debounce time 1 seconds
+    let mut debouncer = new_debouncer(Duration::from_secs(1), None, tx)?;
+
+    // Add a path to be watched. All files and directories at that path and
+    // below will be monitored for changes.
+    debouncer.watcher().watch(&workspace.root_dir, RecursiveMode::Recursive)?;
+
+    let mut screen = std::io::stdout();
+    write!(screen, "{}", termion::cursor::Save).unwrap();
+    screen.flush().unwrap();
+    let _ = compile_workspace_full(workspace, compile_options);
+    for res in rx {
+        let debounced_events = res.map_err(|mut err| err.remove(0))?;
+
+        // We only want to trigger a rebuild if a noir source file has been modified.
+        let noir_files_modified = debounced_events.iter().any(|event| {
+            let mut event_paths = event.event.paths.iter();
+            let event_affects_noir_file =
+                event_paths.any(|path| path.extension().map_or(false, |ext| ext == "nr"));
+
+            matches!(event.kind, EventKind::Modify(_)) && event_affects_noir_file
```

Comment:
I think we want three kinds of events: `Modify`, `Create`, and `Remove`

Reply:
Thanks, deleting files through the IDE would generate `Modify` events but checked now and `rm`ing them wouldn't.